File: zenox/cogs/schedule.py
import datetime
from discord.ext import commands, tasks

from ..bot.bot import Zenox
from ..static.constants import UTC_8
from ..auto_tasks.cleanDB import cleanDB
from ..auto_tasks.topGG import TopGG
from ..auto_tasks.hoyolabCodes import hoyolabCodes
from ..auto_tasks.wikiCodes import wikiCodes
from ..auto_tasks.client_stats import client_stats

class Schedule(commands.Cog):
    def __init__(self, client: Zenox) -> None:
        self.client = client
    
    async def cog_load(self):
        if not self.client.config.schedule:
            return
        self.clean_db.start()
        self.request_hoyolab_codes.start()
        self.request_wiki_codes.start()
        self.publish_wiki_codes.start()
        self.update_topgg_count.start()
        self.update_client_stats.start()
    
    async def cog_unload(self):
        if not self.client.config.schedule:
            return
        self.clean_db.stop()
        self.request_hoyolab_codes.stop()
        self.request_wiki_codes.stop()
        self.publish_wiki_codes.stop()
        self.update_topgg_count.stop()
        self.update_client_stats.stop()
    
    @tasks.loop(time=datetime.time(0, 0, 0, tzinfo=UTC_8))
    async def clean_db(self):
        await cleanDB.execute(self.client)

    @tasks.loop(time=[datetime.time(hour, 0, 0, tzinfo=UTC_8) for hour in [1, 13]])
    async def request_wiki_codes(self):
        await wikiCodes.execute(self.client)
    
    @tasks.loop(time=[datetime.time(hour, 0, 0, tzinfo=UTC_8) for hour in [3, 15]])
    async def publish_wiki_codes(self):
        await wikiCodes.check_publish(self.client)

    @tasks.loop(minutes=3)
    async def request_hoyolab_codes(self):
        await hoyolabCodes.execute(self.client)

    @tasks.loop(time=datetime.time(0, 0, 0, tzinfo=UTC_8))
    async def update_topgg_count(self):
        await TopGG.execute(self.client)
    
    @tasks.loop(time=datetime.time(0, 0, 0, tzinfo=UTC_8))
    async def update_client_stats(self):
        await client_stats.execute(self.client)
    
    @clean_db.before_loop
    @request_hoyolab_codes.before_loop
    @request_wiki_codes.before_loop
    @publish_wiki_codes.before_loop
    @update_topgg_count.before_loop
    @update_client_stats.before_loop
    async def before_loops(self) -> None:
        await self.client.wait_until_ready()

async def setup(client: Zenox) -> None:
    await client.add_cog(Schedule(client))

File: src/cogs/schedule.ts
import { 
  SlashCommandBuilder, 
  CommandInteraction 
} from 'discord.js';
import { Zenox } from '../bot/bot';
import { log } from '../utils/logger';
import { CleanDB } from '../auto_tasks/cleanDB';
import { TopGG } from '../auto_tasks/topGG';
import { HoyolabCodes } from '../auto_tasks/hoyolabCodes';
import { WikiCodes } from '../auto_tasks/wikiCodes';
import { ClientStats } from '../auto_tasks/client_stats';

export class ScheduleCog {
  private client: Zenox;
  public name = 'schedule';
  private intervals: NodeJS.Timeout[] = [];

  constructor(client: Zenox) {
    this.client = client;
    this.setupScheduledTasks();
  }

  private setupScheduledTasks(): void {
    if (!this.client.config.schedule) {
      log.info('Scheduled tasks disabled');
      return;
    }

    log.info('Setting up scheduled tasks...');

    // Clean DB - Daily at midnight UTC+8
    this.scheduleDailyTask('clean_db', 0, 0, () => CleanDB.execute(this.client));

    // Wiki codes - Twice daily at 1:00 and 13:00 UTC+8
    this.scheduleDailyTask('request_wiki_codes', 1, 0, () => WikiCodes.execute(this.client));
    this.scheduleDailyTask('request_wiki_codes', 13, 0, () => WikiCodes.execute(this.client));

    // Publish wiki codes - Twice daily at 3:00 and 15:00 UTC+8
    this.scheduleDailyTask('publish_wiki_codes', 3, 0, () => WikiCodes.checkPublish(this.client));
    this.scheduleDailyTask('publish_wiki_codes', 15, 0, () => WikiCodes.checkPublish(this.client));

    // Hoyolab codes - Every 3 minutes
    this.scheduleIntervalTask('request_hoyolab_codes', 3 * 60 * 1000, () => HoyolabCodes.execute(this.client));

    // Top.gg count - Daily at midnight UTC+8
    this.scheduleDailyTask('update_topgg_count', 0, 0, () => TopGG.execute(this.client));

    // Client stats - Daily at midnight UTC+8
    this.scheduleDailyTask('update_client_stats', 0, 0, () => ClientStats.execute(this.client));

    log.success('All scheduled tasks configured');
  }

  private scheduleDailyTask(name: string, hour: number, minute: number, task: () => Promise<void>): void {
    const interval = setInterval(async () => {
      try {
        log.info(`Executing scheduled task: ${name}`);
        await task();
        log.success(`Completed scheduled task: ${name}`);
      } catch (error) {
        log.error(`Error in scheduled task: ${name}`, error);
      }
    }, this.getTimeUntilNext(hour, minute));

    this.intervals.push(interval);
    log.info(`Scheduled task '${name}' for ${hour.toString().padStart(2, '0')}:${minute.toString().padStart(2, '0')} UTC+8`);
  }

  private scheduleIntervalTask(name: string, intervalMs: number, task: () => Promise<void>): void {
    const interval = setInterval(async () => {
      try {
        log.debug(`Executing interval task: ${name}`);
        await task();
        log.debug(`Completed interval task: ${name}`);
      } catch (error) {
        log.error(`Error in interval task: ${name}`, error);
      }
    }, intervalMs);

    this.intervals.push(interval);
    log.info(`Scheduled interval task '${name}' every ${intervalMs / 1000 / 60} minutes`);
  }

  private getTimeUntilNext(hour: number, minute: number): number {
    const now = new Date();
    const utc8Offset = 8 * 60; // UTC+8 in minutes
    const localOffset = now.getTimezoneOffset();
    const totalOffset = utc8Offset + localOffset;

    const target = new Date();
    target.setHours(hour, minute, 0, 0);
    target.setTime(target.getTime() + totalOffset * 60 * 1000);

    // If target time has passed today, schedule for tomorrow
    if (target <= now) {
      target.setDate(target.getDate() + 1);
    }

    return target.getTime() - now.getTime();
  }

  public data = new SlashCommandBuilder()
    .setName('schedule')
    .setDescription('Schedule management commands');

  public async execute(interaction: CommandInteraction): Promise<void> {
    await interaction.reply({
      content: 'Schedule management module loaded',
      ephemeral: true
    });
  }

  public destroy(): void {
    // Clear all intervals
    for (const interval of this.intervals) {
      clearInterval(interval);
    }
    this.intervals = [];
    
    log.info('All scheduled tasks stopped');
  }

  // Manual task execution methods for testing
  public async executeCleanDB(): Promise<void> {
    log.info('Manually executing clean DB task');
    await CleanDB.execute(this.client);
  }

  public async executeHoyolabCodes(): Promise<void> {
    log.info('Manually executing Hoyolab codes task');
    await HoyolabCodes.execute(this.client);
  }

  public async executeWikiCodes(): Promise<void> {
    log.info('Manually executing wiki codes task');
    await WikiCodes.execute(this.client);
  }

  public async executeTopGG(): Promise<void> {
    log.info('Manually executing Top.gg task');
    await TopGG.execute(this.client);
  }

  public async executeClientStats(): Promise<void> {
    log.info('Manually executing client stats task');
    await ClientStats.execute(this.client);
  }
}